internal/client/githubclient/githubclient.go

```diff
@@ -111,6 +122,139 @@ func (gc *githubClient) GetLatestRelease(ctx context.Context, owner string, repo
+	if err != nil {
+		return nil, fmt.Errorf("unable to list workflow runs: %w", err)
+	}
+
+	if len(runs.WorkflowRuns) == 0 {
+		return nil, nil
+	}
+
+	// runs.WorkflowRuns is sorted by created_at in descending order so the first element is the latest run
+	return &client.WorkflowRun{WorkflowId: *runs.WorkflowRuns[0].WorkflowID, RunId: *runs.WorkflowRuns[0].ID}, nil
+}
+
+func (gc *githubClient) GetWorkflowRunArtifacts(ctx context.Context, owner, repo, githubSBOMName string, runID int64) ([]*client.WorkflowArtifactContent, error) {
+	var res []*client.WorkflowArtifactContent
+
+	// get workflow run artifacts
+	artifacts, _, err := gc.ghClient.Actions.ListWorkflowRunArtifacts(ctx, owner, repo, runID, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list workflow run artifacts: %w", err)
+	}
+	for _, j := range artifacts.Artifacts {
+		// download artifact
+		file, _, err := gc.ghClient.Actions.DownloadArtifact(ctx, owner, repo, j.GetID(), true)
+		if err != nil {
+			return nil, fmt.Errorf("unable to download artifact: %w", err)
+		}
+
+		// If the githubSBOMName is empty, we want to return all artifacts
+		// Otherwise, we only want to return the artifacts that have the name of githubSBOMName
+		if githubSBOMName == "" || (githubSBOMName != "" && *j.Name == githubSBOMName) {
+			arr, err := DownloadAndExtractZip(file.String(), runID)
+			if err != nil {
+				return nil, fmt.Errorf("unable to download and extract zip: %w", err)
+			}
+
+			res = append(res, arr...)
+		}
+	}
+
+	return res, nil
+}
+
+// DownloadAndExtractZip downloads a zip file from the given URL, extracts its contents,
+// and returns only those files that are valid JSON.
+func DownloadAndExtractZip(url string, runID int64) ([]*client.WorkflowArtifactContent, error) {
+	// Download the zip file
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("error getting zip file: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("bad status: %s", resp.Status)
+	}
+
+	// Read the body into a buffer
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %w", err)
+	}
+
+	// Read the zip archive
+	zipReader, err := zip.NewReader(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
+	if err != nil {
+		return nil, fmt.Errorf("error reading zip file: %w", err)
+	}
+
+	var files []*client.WorkflowArtifactContent
+
+	// Iterate through each file in the archive
+	for _, zipFile := range zipReader.File {
+		f, err := zipFile.Open()
+		if err != nil {
+			return nil, fmt.Errorf("error opening file in zip: %w", err)
+		}
+		defer f.Close()
+
+		fileData := make([]byte, zipFile.UncompressedSize64)
+		_, err = io.ReadFull(f, fileData)
+		if err != nil {
+			return nil, fmt.Errorf("error reading file data: %w", err)
+		}
+
+		// Check if the file content is valid JSON
+		if json.Valid(fileData) {
```

Comment:
I'm not sure, but do we need to do this check here? I believe later on in the pipeline `pkg/handler/processor/guesser` will handle the file type and ignore anything it doesn't recognize. For example xml is a valid type as well. Do the other collectors have any filtering? Check the file collector and see what it does. Maybe something to ask about in slack.

Reply:
I have removed the check because https://github.com/guacsec/guac/blob/main/pkg/handler/processor/guesser/format_json.go checks whether it is a valid json.